mgr/dashboard: fix snapshot creation with duplicate name

aaSharma14 opened the pull request:
Snapshot creation with same name on UI throwing 500 Internal Error, This PR intends to fix this issue.

Fixes: https://tracker.ceph.com/issues/57456
Signed-off-by: Aashish Sharma <aasharma@redhat.com>

![Screenshot from 2022-09-05 17-28-04](https://user-images.githubusercontent.com/66050535/188444873-025bd790-ff27-4994-9766-c86d9424b878.png)

![Screenshot from 2022-09-06 10-57-37](https://user-images.githubusercontent.com/66050535/188555346-7fcb7139-0203-4885-bc86-c601c7896deb.png)







<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

epuertat reviewed (approved):
This fixes the UI validation, but what happens if I try this same operation from the API?

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cephfs/cephfs-directories/cephfs-directories.component.ts:
```diff
@@ -546,22 +547,44 @@ export class CephfsDirectoriesComponent implements OnInit, OnChanges {
+    this.selectedDir.snapshots.forEach(name => {
+      snapNames.push(name['name']);
+    });
+    if(snapNames.includes(snapName)) {
+      return this.alreadyExists = true;
+    }
+    else {
+      return this.alreadyExists = false;
+    }
```
?
```suggestion
    return snapNames.includes(snapName);
```

nizamial09 replied:
@aaSharma14 i edited this after realizing my own mistake xD

aaSharma14 replied:
Yeah, I updated this.

aaSharma14 commented:
> This fixes the UI validation, but what happens if I try this same operation from the API?

@epuertat , Added API validation as well

nizamial09 commented:
> > This fixes the UI validation, but what happens if I try this same operation from the API?
> 
> @epuertat , Added API validation as well

@aaSharma14 just a doubt, if we added the API validation, then do we need a frontend validation? I mean the backend will already throw the Exception right and it'll be caught and shown in the frontend as well. Maybe an async validator in the form name field will be good IMHO.

pereman2 commented:
jenkins test make check

epuertat commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cephfs/cephfs-directories/cephfs-directories.component.ts:
```diff
@@ -546,22 +547,39 @@ export class CephfsDirectoriesComponent implements OnInit, OnChanges {
   }
 
+  validateValue(control: AbstractControl) {
+    const snapName = control.value;
+    let snapNames: string[] = [];
+    this.selectedDir.snapshots.forEach(name => {
+      snapNames.push(name['name']);
+    });
+    this.alreadyExists = snapNames.includes(snapName);
```
```suggestion
    this.alreadyExists = this.selectedDir.snapshots.some(s => s.name == control.value);
```

aaSharma14 commented:
> > > This fixes the UI validation, but what happens if I try this same operation from the API?
> > 
> > 
> > @epuertat , Added API validation as well
> 
> @aaSharma14 just a doubt, if we added the API validation, then do we need a frontend validation? I mean the backend will already throw the Exception right and it'll be caught and shown in the frontend as well. Maybe an async validator in the form name field will be good IMHO.

@nizamial09 , I think by adding the validation to the frontend as well will prevent sending the API request at first place

aaSharma14 commented:
> > > This fixes the UI validation, but what happens if I try this same operation from the API?
> > 
> > 
> > @epuertat , Added API validation as well
> 
> @aaSharma14 just a doubt, if we added the API validation, then do we need a frontend validation? I mean the backend will already throw the Exception right and it'll be caught and shown in the frontend as well. Maybe an async validator in the form name field will be good IMHO.

@nizamial09 , I think by adding the validation to the frontend as well will prevent sending the API request at first place

aaSharma14 commented:
jenkins test dashboard cephadm

aaSharma14 commented:
jenkins test windows

aaSharma14 commented:
jenkins test dashboard cephadm

aaSharma14 commented:
jenkins test dashboard cephadm
